Documentation/PhpArchitecture/Readonly.rst

```diff
@@ -0,0 +1,131 @@
+..  include:: /Includes.rst.txt
+..  index:: pair: Coding guidelines; About readonly
+..  _php-architecture-readonly:
+
+=====================
+About :php:`readonly`
+=====================
+
+PHP v8.1 introduced
+`readonly properties <https://www.php.net/manual/en/language.oop5.properties.php#language.oop5.properties.readonly-properties>`_
+while PHP v8.2 added
+`readonly classes <https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.readonly>`_.
+:php:`readonly` properties can only be written once - usually in the constructor.
+
+Declaring :ref:`services <cgl-services>` and
+:ref:`value objects <cgl-named-arguments-pcpp-value-objects>` as readonly is
+beneficial for TYPO3 Core and extensions, offering immutability and clarity
+regarding the statelessness of services.
+
+This document discusses the use of readonly within the TYPO3 Core ecosystem,
+outlining best practices for TYPO3 extension and Core developers regarding the
+adoption and avoidance of this language feature.
+
+
+Readonly services
+-----------------
+
+Readonly properties align seamlessly with services using
+:ref:`constructor injection <Constructor-injection>`, e.g.:
+
+.. code-block:: php
+
+    final class UserController
+    {
+        private string $someProperty = 'foo';
+
+        public function __construct(
+            private readonly SomeDependency $someDependency,
+        ) {}
+
+        // ...
+    }
+
+Well designed stateless services with no properties apart from those declared using
+`constructor property promotion <https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion>`
+can be declared :php:`readonly` on class level:
+
+.. code-block:: php
+
+    final readonly class UserController
+    {
+        public function __construct(
+            private SomeDependency $someDependency,
+        ) {}
+
+        // ...
+    }
+
+Declaring properties or - even better - entire service classes readonly is a great
+way to clarify possible impact of state within services: If used correctly, readonly
+tells developers this service is stateless, shared, and can be used in any
+context and as often as needed without side effects from previous usages, and without
+influencing possible further usages within the same request. Statelessness
+is an important asset of services and readonly helps to sort out this question.
+
+Even when a service class is declared as readonly, ensuring immutability at its level,
```

Comment:
I miss a sentence that properties of a stateless object which is inserted into another object as readonly property can still be modified.

Reply:
Well, if I understand you correctly, that's basically the opposite of what is stated. I'm unsure on how to phrase this in a way that it merges well into the existing paragraphs and tend to not add something like this. I hope that's ok.